hydroflow_lang/src/graph/mod.rs

```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
-                    subgraph_send_handoffs[node_subgraph[src]].push(dst);
-                }
-                (Node::Handoff, Node::Operator(_)) => {
-                    subgraph_recv_handoffs[node_subgraph[dst]].push(src);
+        let mut components = Default::default();
+        for sg in stack.into_iter().rev() {
+            assign(&subgraph_preds, sg, sg, &mut components);
+        }
+
+        components
+    };
+
+    let topo_sort_order = {
+        // Condensed each SCC into a single node for toposort.
+        let condensed_preds: HashMap<GraphSubgraphId, Vec<GraphSubgraphId>> = subgraph_preds
+            .iter()
+            .map(|(u, preds)| (scc[u], preds.iter().map(|v| scc[v]).collect()))
+            .collect();
+
+        graph_algorithms::topo_sort(subgraph_nodes.keys(), |v| {
+            condensed_preds.get(&v).into_iter().flatten().cloned()
+        })
+    };
+
+    let mut subgraph_stratum: SecondaryMap<GraphSubgraphId, usize> = topo_sort_order
+        .into_iter()
+        .enumerate()
+        .map(|(n, sg_id)| (sg_id, n))
+        .collect();
+
+    let max_stratum = subgraph_stratum.values().cloned().max().unwrap_or(0) + 1;
+    for &(src, dst, dst_idx, input_barrier) in barrier_crossers.iter() {
+        let src_sg = node_subgraph[src];
+        let dst_sg = node_subgraph[dst];
+        let src_stratum = subgraph_stratum[src_sg];
+        let dst_stratum = subgraph_stratum[dst_sg];
+        match input_barrier {
+            InputBarrier::None => unreachable!("No barrier"),
+            InputBarrier::Stratum => {
+                // Any negative edges which go onto the same or previous stratum are bad.
+                // Indicates an unbroken negative cycle.
+                if dst_stratum <= src_stratum {
+                    dst_idx
+                        .span()
+                        .unwrap()
+                        .error("Connection creates a negative cycle which must be broken with a `next_epoch()` operator.")
+                        .emit();
+                    return Err(());
                 }
-                (Node::Handoff, Node::Handoff) => {
-                    Span::call_site().unwrap().error(format!(
-                        "Internal Error: Consecutive handoffs {:?} -> {:?}",
-                        src.data(),
-                        dst.data()
-                    ));
+            }
+            InputBarrier::Epoch => {
+                // If epoch edge goes foreward, need to buffer. (TODO: use a different kind of handoff).
+                if src_stratum <= dst_stratum {
+                    // Need to inject.
+                    let new_node_id = nodes.insert(Node::Operator(parse_quote! { identity() }));
+                    let new_subgraph_id = subgraph_nodes.insert(vec![new_node_id]);
+                    subgraph_stratum.insert(new_subgraph_id, max_stratum);
+                    node_subgraph.insert(new_node_id, new_subgraph_id);
```

Comment:
It's kind of a bummer to have to remember all the SecondaryMaps to be handled everywhere. Can we refactor into a single helper?